Sqlite in cloud documentation (#740)
fermyon/developer: content/cloud/noops-sql-db.md

Reviewer: Video is scheduled to publish at the same time as this page Noon 26th July (UTC).

```

>> If you need to look up the name of your database, use the `spin cloud sqlite list` command

### Manage NoOps SQL Database Lifecycle

A NoOps SQL Database is created when you deploy a Spin application that references a "default" SQLite database in its component manifest. Multiple components in the same Spin application can be given access to the default database like so:

```toml
# c1 does not have access to the default database
[component]
name = "c1"

# c2 can use the default database
[component]
name = "c2"
sqlite_databases = ["default"]

# c3 can use default database as well
[component]
name = "c3"
sqlite_databases = ["default"]
```

The NoOps SQL database will created when the Spin application is deployed. The data persists Spin application invocations and updates. 

```bash
$ spin cloud deploy
Deploying app ..
Creating database inspirational-pig (default) for app "quickstart"
```

You can check to see your NoOps SQL database by using the `spin cloud sqlite` command:

>> Fermyon Cloud will automatically generate a random name for your NoOps SQL database.

```bash
$ spin cloud sqlite list
inspirational-pig (default) 
```

> Lastly, after [deleting your Spin application](/cloud/delete.md) you can delete your default database with the following command:

Reviewer: ```suggestion
Lastly, (warning; using the `delete` subcommand is **permanent**) you can delete your default database with the following command:
```

The action is irreversible.
Please type "strong-boson" for confirmation: strong-boson
Deleting database ...
Database "strong-boson" deleted
```

> In the Private Beta, the database is inaccessible once the application has been deleted. However, we separate out the the database deletion step because we anticipate the database being accessible after deletion in a future beta. We enforce database deletion to prevent unintentional data sharing between Spin applications. 

## FAQ

- **Can I export data from my database?**
At this time, Fermyon Cloud does not have a data export feature. We'd love to hear your thoughts on what an ideal experience would look like at [github.com/fermyon/feedback](https://github.com/fermyon/feedback)

- **If I upgrade my Spin application, can I still access the same instance of my database?**
Yes, as long as you haven't deleted your database, the instance will be running and stateful. 

- **If I delete and redeploy my Spin application, can I still access the same instance of my database?**
No, at this time you cannot access the database. You will need to delete the instance and provision a new one. 

## Next Steps

> * For support or feature feedback, please join our [Discord channel #noops-sql-db-private-beta](https://www.fermyon.com/discord). This is where you can provide feedback and ask for assistance from the Fermyon Cloud team on all things NoOps SQL Database. 

Author: ```suggestion
* For support or feature feedback, please join our [Discord channel #no-ops-db-beta](https://discord.gg/XxkZvsya). This is where you can provide feedback and ask for assistance from the Fermyon Cloud team on all things NoOps SQL Database. 
```

title = "NoOps SQL Database"
template = "cloud_main"

> date = "2023-07-26T16:00:00Z"

Reviewer: ```suggestion
date = "2023-07-26T12:00:00Z"
```

enable_shortcodes = false
[extra]
url = "https://github.com/fermyon/developer/blob/main/content/cloud/noops-sql-db.md"

---

- [Accessing Private Beta](#accessing-private-beta)
- [Service Limitations and Quotas For NoOps SQL Database in Fermyon Cloud](#service-limitations-and-quotas-for-noops-sql-database-in-fermyon-cloud)
- [How To:](#how-to)
	- [Prepare a NoOps SQL Database](#prepare-a-noops-sql-database)
	- [Manage NoOps SQL Database Lifecycle](#manage-noops-sql-database-lifecycle)
- [FAQ](#faq)
- [Next Steps](#next-steps)

> With NoOps SQL Database using SQLite support in Fermyon Cloud, you can persist relational data generated by your Spin application in between application invocations without having to worry about database management. Fermyon Cloud will provision and manage the database on your behalf, no operations (NoOps) required. Fermyon Cloud will do the heavy lifting for you. To learn more about the Spin SQLite SDK, please visit the [API guide](../spin/sqlite-api-guide.md).

Reviewer: ```suggestion
With NoOps SQL Database using SQLite support in Fermyon Cloud, you can persist relational data generated by your Spin application in between application invocations without having to worry about database management. Fermyon Cloud will provision and manage the database on your behalf, no operations (NoOps) required. Fermyon Cloud will do the heavy lifting for you. To learn more about the Spin SQLite SDK, please visit the [API guide](../spin/sqlite-api-guide.md).

<iframe width="854" height="480" src="https://www.youtube.com/embed/buwr66oRPbU" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>

```

## Accessing Private Beta

NoOps SQL Database in Fermyon Cloud is currently in private beta. To request access to the private beta, please fill out this short [sign-up form](https://fibsu0jcu2g.typeform.com/to/Brv12FI0#hubspot_utk=xxxxx&hubspot_page_name=xxxxx&hubspot_page_url=xxxxx).
 
> Please note that the private beta is limited in space and all requests cannot be guaranteed. 

## Quotas And Service Limitations For NoOps SQL Database in Fermyon Cloud

*Quotas* 
* You can have a maximum of 1 NoOps SQL Database in your Cloud account
* Your NoOps SQL Database can hold up 1 GB of data

*Service Limitations*
* You must delete your Spin application before you can delete your NoOps SQL Database
* The `--sqlite` flag not supported on `spin cloud deploy`
* Fermyon Cloud only supports "default" NoOps SQL Databases at this time

## How To:

### Prepare a NoOps SQL Database

> We recommend using the `execute` command in your application's source code to create a table. For example, in this routing function we use the Rust SDK to check if the table exists, and to create it if not:

Reviewer 2: The `spin cloud sqlite execute` command should be live by the time private preview starts. Should we remove this?

sqlite_databases = ["default"]

# c3 can use default database as well
[component]
name = "c3"
sqlite_databases = ["default"]
```

The NoOps SQL database will created when the Spin application is deployed. The data persists Spin application invocations and updates. 

```bash
$ spin cloud deploy
Deploying app ..
Creating database inspirational-pig (default) for app "quickstart"
```

You can check to see your NoOps SQL database by using the `spin cloud sqlite` command:

>> Fermyon Cloud will automatically generate a random name for your NoOps SQL database.

```bash
$ spin cloud sqlite list
inspirational-pig (default) 
```

Lastly, after [deleting your Spin application](/cloud/delete.md) you can delete your default database with the following command:

```bash
$ spin cloud sqlite delete inspirational-pig
inspirational-pig has been deleted
```

> In the Private Beta, the database is inaccessible once the application has been deleted. However, we separate out the the database deletion step because we anticipate the database being accessible after deletion in a future beta. We enforce NoOps SQL Database deletion to prevent unintentional data sharing between Spin applications. 

> ## FAQ

Reviewer 3: Somewhere (whether here or elsewhere on the page) it ought to explain "if you do a deploy and hit such-and-such an error, it means the app wants a database but the one database slot is already taken up"

> title = "NoOps SQL Database"

Reviewer 2: Can we refer to this feature as "NoOps SQLite Database"? It's important to note that it is a sqlite database.

<iframe width="854" height="480" src="https://www.youtube.com/embed/buwr66oRPbU" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>

## Accessing Private Beta

NoOps SQL Database in Fermyon Cloud is currently in private beta. To request access to the private beta, please fill out this short [sign-up form](https://fibsu0jcu2g.typeform.com/to/Brv12FI0#hubspot_utk=xxxxx&hubspot_page_name=xxxxx&hubspot_page_url=xxxxx).
 
> Please note that the private beta is limited in space and all requests cannot be guaranteed. 

## Quotas And Service Limitations For NoOps SQL Database in Fermyon Cloud

*Quotas* 
* You can have a maximum of 1 NoOps SQL Database in your Cloud account
* Your NoOps SQL Database can hold up 1 GB of data

*Service Limitations*

> * You must delete your Spin application before you can delete your NoOps SQL Database

Reviewer 2: This isn't currently true. Do we want this to be true?

Author: Existing functionality is OK. The database will be stranded either way after the application is deleted, so I don't think the order of operations is important. I will remove the service limitation from the developer docs.

```rust
#[http_component]
fn handle_todo(req: Request) -> anyhow::Result<Response> {
    let connection = Connection::open_default()?;
		connection.execute("CREATE TABLE IF NOT EXISTS todos (
			  id INTEGER PRIMARY KEY AUTOINCREMENT,
			  description TEXT NOT NULL,
			  due_date DATE,
			  starred BOOLEAN DEFAULT 0,
			  is_completed BOOLEAN DEFAULT 0
				);", &[])?;

    let router = http_router! {
<...>
    };
    router.handle(req)
}
```

### Manage NoOps SQL Database Lifecycle

A NoOps SQL Database is created when you deploy a Spin application that references a "default" SQLite database in it's component manifest. Multiple components in the same Spin application can be given access to the default database like so:

```toml
# c1 does not have access to the default database
[component]
name = "c1"

# c2 can use the default database
[component]
name = "c2"
sqlite_databases = ["default"]

# c3 can use default database as well
[component]
name = "c3"
sqlite_databases = ["default"]
```

> The NoOps SQL database will created when the Spin application is deployed. The data in between Spin application invocations and updates. 

Reviewer 4: ```suggestion
The NoOps SQL database will created when the Spin application is deployed. The data persists Spin application invocations and updates. 
```

```bash
$ spin cloud deploy
Deploying app ..
Creating database inspirational-pig (default) for app "quickstart"
```

> You can check to see which Spin application is tied to your NoOpsSQL database by using the `spin cloud sql` command:

Reviewer 4: ```suggestion
You can check to see which Spin application is tied to your NoOpsSQL database by using the `spin cloud sqlite` command:
```

>> Fermyon Cloud will automatically generate a random name for your NoOps SQL database.

```bash
$ spin cloud sqlite list

> inspirational-pig (default) (currenly used by "quickstart")

Reviewer 4: ```suggestion
inspirational-pig (default)
```

```

Lastly, after [deleting your Spin application](/cloud/delete.md) you can delete your default database with the following command:

```bash

> $ spin cloud sqlite delete default

Reviewer 4: ```suggestion
$ spin cloud sqlite delete inspirational-pig
```

> inspirational-pig has been deleted

Reviewer 4: I am not sure what output we have on deletes currently. this may need to be updated... or the cli updated to match this

Reviewer 2: The output has been updated, so this should be updated.

```

> > The NoOps SQL Database must be deleted before another Spin application can use the feature to prevent unintentional data sharing between Spin applications. 

Reviewer 3: Please could we say "NoOps SQL Database" less often?  After a while it feels a bit artificial (at least to me).

Author: Its a bit tricky as its the feature name. Kate left similar feedback. I wonder if there's a shorthand we can use to avoid name fatigue.

Reviewer: Just checking ... We use it 21 times; any suggestions on what to use instead?

Reviewer: Done in new commit. From 21 down to 9 :)

## FAQ

- **Can I export data from my NoOps SQL Database?**
At this time, Fermyon Cloud does not have a data export feature for NoOps SQL DB. We'd love to hear your thoughts on what an ideal experience would look like at [github.com/fermyon/feedback](https://github.com/fermyon/feedback)

- **If I upgrade my Spin application, can I still access the same instance of my NoOps SQL Database?**
Yes, as long as you haven't deleted your NoOps SQL Database, the instance will be running and stateful. 

- **If I delete and redeploy my Spin application, can I still access the same instance of my NoOps SQL Database?**

> TBD!!!

Reviewer 4: No

## Next Steps

> * For support or feature feedback, please join our [Discord channel #noops-sql-db-private-beta](LINK-TBD)

Reviewer 3: @macolso This needs defining!

> title = "NoOps SQL Database"

Reviewer: ```suggestion
title = "NoOps SQLite Database"
```

<iframe width="854" height="480" src="https://www.youtube.com/embed/buwr66oRPbU" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>

## Accessing Private Beta

NoOps SQL Database in Fermyon Cloud is currently in private beta. To request access to the private beta, please fill out this short [sign-up form](https://fibsu0jcu2g.typeform.com/to/Brv12FI0#hubspot_utk=xxxxx&hubspot_page_name=xxxxx&hubspot_page_url=xxxxx).
 
> Please note that the private beta is limited in space and all requests cannot be guaranteed. 

## Quotas And Service Limitations For NoOps SQL Database in Fermyon Cloud

*Quotas* 
* You can have a maximum of 1 NoOps SQL Database in your Cloud account
* Your NoOps SQL Database can hold up 1 GB of data

*Service Limitations*
* The `--sqlite` flag not supported on `spin cloud deploy`

> * Fermyon Cloud only supports "default" NoOps SQL Databases at this time

Reviewer 2: I'm not sure if this will make sense to everyone. Perhaps:

```suggestion
* Fermyon Cloud only supports NoOps SQLite Databases named "default" at this time
```

## How To:

### Prepare a NoOps SQL Database

You can use the `spin cloud sqlite execute` command to create a table in your database. 

> 

Reviewer 2: Should we give an example? It seems we're giving much more weight to the alternative approach (which I personally think is a bad idea). 

```bash
spin cloud sqlite execute confident-plum "CREATE TABLE IF NOT EXISTS todos (id INTEGER PRIMARY KEY AUTOINCREMENT, description TEXT NOT NULL);"
```

In fact, I would argue we remove the second approach entirely from the docs...

### Manage NoOps SQL Database Lifecycle

A NoOps SQL Database is created when you deploy a Spin application that references a "default" SQLite database in its component manifest. Multiple components in the same Spin application can be given access to the default database like so:

```toml
# c1 does not have access to the default database
[component]
name = "c1"

# c2 can use the default database
[component]
name = "c2"
sqlite_databases = ["default"]

# c3 can use default database as well
[component]
name = "c3"
sqlite_databases = ["default"]
```

The NoOps SQL database will created when the Spin application is deployed. The data persists Spin application invocations and updates. 

```bash
$ spin cloud deploy
Deploying app ..
Creating database inspirational-pig (default) for app "quickstart"
```

You can check to see your NoOps SQL database by using the `spin cloud sqlite` command:

>> Fermyon Cloud will automatically generate a random name for your NoOps SQL database.

```bash
$ spin cloud sqlite list
inspirational-pig (default) 
```

> Lastly, after [deleting your Spin application](/cloud/delete.md) you can delete your default database with the following command:

Reviewer: Thanks @rylev

The action is irreversible.
Please type "strong-boson" for confirmation: strong-boson
Deleting database ...
Database "strong-boson" deleted
```

> In the Private Beta, the database is inaccessible once the application has been deleted. However, we separate out the the database deletion step because we anticipate the database being accessible after deletion in a future beta. We enforce NoOps SQL Database deletion to prevent unintentional data sharing between Spin applications. 

## FAQ

- **Can I export data from my NoOps SQL Database?**
At this time, Fermyon Cloud does not have a data export feature. We'd love to hear your thoughts on what an ideal experience would look like at [github.com/fermyon/feedback](https://github.com/fermyon/feedback)

- **If I upgrade my Spin application, can I still access the same instance of my NoOps SQL Database?**
Yes, as long as you haven't deleted your database, the instance will be running and stateful. 

- **If I delete and redeploy my Spin application, can I still access the same instance of my NoOps SQL Database?**
No, at this time you cannot access the database. You will need to delete the instance and provision a new one. 

## Next Steps

> * For support or feature feedback, please join our [Discord channel #noops-sql-db-private-beta](LINK-TBD). This is where you can provide feedback and ask for assistance from the Fermyon Cloud team on all things NoOps SQL Database. 

Reviewer: ```suggestion
* For support or feature feedback, please join our [Discord channel #noops-sql-db-private-beta](https://www.fermyon.com/discord). This is where you can provide feedback and ask for assistance from the Fermyon Cloud team on all things NoOps SQL Database. 
```

title = "NoOps SQL Database"
template = "cloud_main"

> date = "2023-07-26T12:00:00Z"

Reviewer: ```suggestion
date = "2023-07-26T16:00:00Z"
```

---

- [Accessing Private Beta](#accessing-private-beta)
- [Service Limitations and Quotas For NoOps SQL Database in Fermyon Cloud](#service-limitations-and-quotas-for-noops-sql-database-in-fermyon-cloud)
- [How To:](#how-to)
	- [Prepare a NoOps SQL Database](#prepare-a-noops-sql-database)
	- [Manage NoOps SQL Database Lifecycle](#manage-noops-sql-database-lifecycle)
- [FAQ](#faq)
- [Next Steps](#next-steps)

With NoOps SQL Database using SQLite support in Fermyon Cloud, you can persist relational data generated by your Spin application in between application invocations without having to worry about database management. Fermyon Cloud will provision and manage the database on your behalf, no operations (NoOps) required. Fermyon Cloud will do the heavy lifting for you. To learn more about the Spin SQLite SDK, please visit the [API guide](../spin/sqlite-api-guide.md).

<iframe width="854" height="480" src="https://www.youtube.com/embed/buwr66oRPbU" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>

> 

Reviewer: ```suggestion
```

## Accessing Private Beta

NoOps SQL Database in Fermyon Cloud is currently in private beta. To request access to the private beta, please fill out this short [sign-up form](https://fibsu0jcu2g.typeform.com/to/Brv12FI0#hubspot_utk=xxxxx&hubspot_page_name=xxxxx&hubspot_page_url=xxxxx).
 
> Please note that the private beta is limited in space and all requests cannot be guaranteed. 

## Quotas And Service Limitations For NoOps SQL Database in Fermyon Cloud

*Quotas* 
* You can have a maximum of 1 NoOps SQL Database in your Cloud account
* Your NoOps SQL Database can hold up 1 GB of data

*Service Limitations*
* The `--sqlite` flag not supported on `spin cloud deploy`
* Fermyon Cloud only supports "default" NoOps SQL Databases at this time

## How To:

### Prepare a NoOps SQL Database

You can use the `spin cloud sqlite execute` command to create a table in your database. 

> 

Reviewer: ```suggestion
```

Alternatively, you can use `execute` command in your application's source code as well. For example, in this routing function we use the Rust SDK to check if the table exists, and to create it if not:

```rust
#[http_component]
fn handle_todo(req: Request) -> anyhow::Result<Response> {
    let connection = Connection::open_default()?;
	connection.execute("CREATE TABLE IF NOT EXISTS todos (
			id INTEGER PRIMARY KEY AUTOINCREMENT,
			description TEXT NOT NULL,
			due_date DATE,
			starred BOOLEAN DEFAULT 0,
			is_completed BOOLEAN DEFAULT 0
			);", &[])?;
<...>
}
```

> As pointed out in the [API guide](/spin/sqlite-api-guide.md), this strategy can end up mingling your “hot path” application logic with database maintenance code; decide which approach is best based on your application’s needs.

Reviewer: ```suggestion

As pointed out in the [API guide](/spin/sqlite-api-guide.md), this strategy can end up mingling your “hot path” application logic with database maintenance code; decide which approach is best based on your application’s needs.
```

Reviewer: Need line spacing for CI to pass tests.
